warl_proposal/warl_proposal.rst

```diff
@@ -0,0 +1,180 @@
+###################
+WARL field Proposal
+###################
+
+Since the RISC-V privilege spec indicates several CSRs and sub-fields of CSRs to be WARL (Write-Any-Read-Legal), it is now necessary to provide a scheme of WARL functions which can be used to precisely define the functionality of any such WARL field/register. This document provides a proposal to generalize these definitions.
+
+**Authors**: Allen, Pawan, Neel
+
+
+WARL-FUNC(*warl-func*)
+======================
+
+The WARL function can be defined in two ways:
+    - a list of single-value-tuples explicitly specifying the distinct legal values that the field/register can take. 
+    - a list of four-value-tuples defining a specific range of legal values that the field/register can take.
+
+1. Single-Value-Tuple Description:
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+These can simply be enumerated as [ [0], [3], [19], ...]. 
+This means that the respective csr can only take legal values: 0, 3, 19, ....
+
+2. Four-Value-Tuple Description:
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+  Here the legal values of the field/register is defined using 4 fields:
+
+  .. code-block:: python
+
+    [ mask, fixedval, lower, upper]
+
+
+  - **mask** : A bit with value 1 indicates that the bit is writable. A bit value of 0  indicates the bit is hardwired to a value corresponding to the same bit in the fixedval field.  
+  - **fixedval** : For all the bits that are 0s in mask, the corresponding bits in this field indicate the hardwired value (can be 0 or 1).
+  - **lower**: This defines the lower-bound of legal values.
+  - **upper**: This defines the upper-bound of legal values.
+
+  The *base* and *bound* values are defined as :
+  
+  .. code-block:: python
+
+      base  = (lower & mask) | (fixedval & ~mask)
+      bound = (upper & mask) | (fixedval & ~mask)
+
+  The value (val) to be written to the csr/register is defined as:
+  
+  .. code-block:: python
+
+      val = (write-val & mask) | (fixedval & ~mask)
+
+  This value is legal only if it lies within the base and bound values:
+
+  .. code-block:: python
+   
+    base <= val <= bound
+
+
+  Now if val is an illegal value, the field/register should resort to a known legal value. There are multiple options possible and they defined using the **mode** field which can be any of the following:
```

Comment:
The mode seems to quite complicated, how is it used in real application?

Reply:
So this piece of proposal was given by Allen (from esperanto) and it sort of tries to cover all possible scenarios one "could" come up with. I agree the general case is typically going to be "unchanged". 
But for example in spike the MPP field in mstatus takes the highest privilege when certain bits in misa are not set. This would thus use the "largest" mode when illegal values are encountered.

Reply:
I feel that instead of enumerating all known conditions, it's better to design some simple specification syntax which can be used here and other places. Things like nearup, neardown, nextup, nextdown doesn't seem to be very easy to understand to me. And there could be some very custom relationship which is not in the list.  It could be something more explicit:
```
wr val == 0x50 -> unchanged
wr val == 0x40 -> rd val == 0x41
```

Reply:
taken care of this in the new draft.